environments/centos7-base/main.py

```diff
@@ -0,0 +1,204 @@
+#!/usr/bin/env python3
+
+import os
+import sys
+import stat
+import json
+import urllib2
+
+from datetime import datetime
+from subprocess import call
+from string import Template
+
+def error(message):
+    log(Template('ERR: $message').substitute(message=message))
+    sys.stdout.flush()
+
+def warn(message):
+    log(Template('WARN: $message').substitute(message=message))
+    sys.stdout.flush()
+
+def log(message):
+    print(Template('[main] $message').substitute(message=message))
+    sys.stdout.flush()
+
+# This function will attempt to read any results file, add our metadata to it,
+# write it to disk, send everything to S3, and exit the program.
+#
+# If succeeded is false, we'll assume things failed badly enough that we shouldn't
+# even try to read /grade/results/results.json.
+def finish(succeeded, info):
+    if info['job_id'] is None:
+        # Someone screwed up bad; without a job ID, we can't upload archives to
+        # S3 or notify the webhook that we've failed
+        error('job_id was not specified, so it\'s impossible to upload an archive')
+        sys.exit(1)
+
+    data = None
+    if succeeded and os.path.isfile('/grade/results/results.json'):
+        # Let's try to read their results file
+        # They successfully wrote out the results file!
+        with open('/grade/results/results.json') as json_data:
+            try:
+                data = json.load(json_data)
+            except JSONDecodeError:
+                data = None
+
+    results = {}
+    results['data'] = data
+    results['start_time'] = info['start_time']
+    results['end_time'] = datetime.utcnow().isoformat()
+
+    # Note the distinction between grade/results/results.json and grade/results.json
+    # The former is generated by the code that we're calling and will store
+    # test results; the latter will hold the results file that we'll generate
+    # that contains both test results and some addition metadata
+    with open('grade/results.json', 'w+') as results_file:
+        json.dump(results, results_file)
+
+    # Now we can upload this results file to AWS
+    if info['results_bucket']:
+        s3_results_file = Template('s3://$bucket/job_$job.json').substitute(bucket=info['results_bucket'], job=info['job_id'])
+        s3_results_push_ret = call(['aws', 's3', 'cp', '/grade/results.json', s3_results_file])
+        if s3_results_push_ret is not 0:
+            error('could not push results to S3')
+    else:
+        error('the results bucket was not specified')
+
+    # Now let's make an archive of the /grade directory and send it back to S3
+    # for storage
+    if info['archives_bucket']:
+        zip_ret = call(['tar', '-zcf', '/archive.tar.gz', '/grade/'])
+        if zip_ret is not 0:
```

Comment:
Normally we use `==` for value comparisons like this, and only use `is`/`is not` for identity comparison.